pkg/plugins/runtime/k8s/webhooks/injector/injector.go

```diff
@@ -80,36 +83,106 @@ func (i *KumaInjector) InjectKuma(pod *kube_core.Pod) error {
 		if pod.Spec.InitContainers == nil {
 			pod.Spec.InitContainers = []kube_core.Container{}
 		}
-		pod.Spec.InitContainers = append(pod.Spec.InitContainers, i.NewInitContainer(pod))
+		ic, err := i.NewInitContainer(pod)
+		if err != nil {
+			return err
+		}
+		pod.Spec.InitContainers = append(pod.Spec.InitContainers, ic)
+	}
+
+	if err := i.reconcileProbes(pod); err != nil {
+		return err
 	}
 
 	return nil
 }
 
-func needInject(pod *kube_core.Pod, ns *kube_core.Namespace) (bool, error) {
-	isEnabled := func(enabled string) (bool, error) {
-		switch enabled {
-		case metadata.KumaSidecarInjectionEnabled:
-			return true, nil
-		case metadata.KumaSidecarInjectionDisabled:
-			return false, nil
+func (i *KumaInjector) reconcileProbes(pod *kube_core.Pod) error {
+	if ok, err := needVirtualProbes(pod, i.cfg); err != nil || !ok {
+		return err
+	}
+
+	port, err := virtualProbesPort(pod, i.cfg)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range pod.Spec.Containers {
+		if c.LivenessProbe != nil && c.LivenessProbe.HTTPGet != nil {
+			if err := reconcileProbe(c.LivenessProbe, port); err != nil {
+				return err
+			}
+		}
+		if c.ReadinessProbe != nil && c.ReadinessProbe.HTTPGet != nil {
+			if err := reconcileProbe(c.ReadinessProbe, port); err != nil {
+				return err
+			}
 		}
-		return false, errors.Errorf("annotation %s has wrong value", metadata.KumaSidecarInjectionAnnotation)
 	}
-	if enabled, ok := pod.Annotations[metadata.KumaSidecarInjectionAnnotation]; ok {
-		return isEnabled(enabled)
+	return nil
+}
+
+func needVirtualProbes(pod *kube_core.Pod, cfg runtime_k8s.Injector) (bool, error) {
+	enabled, exist, err := metadata.Annotations(pod.Annotations).GetEnabled(metadata.KumaVirtualProbesAnnotation)
+	if err != nil {
+		return false, err
 	}
-	if enabled, ok := ns.Annotations[metadata.KumaSidecarInjectionAnnotation]; ok {
-		return isEnabled(enabled)
+	if exist {
+		return enabled, nil
+	}
+	return cfg.VirtualProbesEnabled, nil
+}
+
+func virtualProbesPort(pod *kube_core.Pod, cfg runtime_k8s.Injector) (uint32, error) {
+	port, exist, err := metadata.Annotations(pod.Annotations).GetUint32(metadata.KumaVirtualProbesPortAnnotation)
+	if err != nil {
+		return 0, err
+	}
+	if exist {
+		return port, nil
+	}
+	return cfg.VirtualProbesPort, nil
+}
+
+func reconcileProbe(probe *kube_core.Probe, probePort uint32) error {
+	if inbound, ok := probes.KumaProbe(*probe).ToInbound(probePort); ok {
+		// we are dealing with a probe which is already virtual,
+		// all virtual probes should be presented in the dataplane
+		if inbound.Port() == probePort {
+			return errors.Errorf("can't reconcile Pod's probes, inbound port can't be equal to %d,"+
+				" it is reserved value for insecure probes port", probePort)
+		}
+		return nil
+	}
+	virtual := probes.KumaProbe(*probe).ToVirtual(probePort)
+	probe.HTTPGet.Port = intstr.FromInt(int(virtual.Port()))
+	probe.HTTPGet.Path = virtual.Path()
+	return nil
+}
```

Comment:
How about moving all stuff connected with probes to `probes.go` in this package?